src/util/bson/bson.cpp

```diff
@@ -520,106 +523,64 @@ class Parser : public nlohmann::json_sax<json> {
 
-        BsonContainer(BsonArray&& v) noexcept : m_type(ARRAY), array(new BsonArray(std::move(v))) {}
+        BsonContainer(BsonArray&& v) noexcept :
+            m_type(ARRAY),
+            array(new BsonArray(std::move(v))) {}
 
-        explicit operator const BsonDocument&&() const {
+        explicit operator const BsonDocument&() noexcept {
             REALM_ASSERT(is_document());
-            return std::move(*document);
+            return *document;
         }
 
-        explicit operator const BsonArray&&() const {
+        explicit operator const BsonArray&() const noexcept {
             REALM_ASSERT(is_array());
-            return std::move(*array);
+            return *array;
         }
 
-        BsonContainer(const BsonContainer& v)
-        {
-            m_type = v.m_type;
-            if (m_type == DOCUMENT) {
-                document = new BsonDocument;
-                *document = *v.document;
-            } else {
-                array = new BsonArray;
-                *array = *v.array;
-            }
-        }
-
-        BsonContainer& operator=(const BsonContainer& v)
-        {
-            if (&v == this) return *this;
-            this->~BsonContainer();
-
-            m_type = v.m_type;
-            if (m_type == DOCUMENT) {
-                document = new BsonDocument;
-                *document = *v.document;
-            } else {
-                array = new BsonArray;
-                *array = *v.array;
-            }
-            return *this;
-        }
-
-        BsonContainer& operator=(BsonContainer&& v)
-        {
-            if (&v == this) return *this;
-            this->~BsonContainer();
-
-            m_type = v.m_type;
-            if (m_type == DOCUMENT) {
-                document = v.document;
-                v.document = nullptr;
-            } else {
-                array = v.array;
-                v.array = nullptr;
-            }
-            return *this;
+        explicit operator BsonDocument&() noexcept {
+            REALM_ASSERT(is_document());
+            return *document.release();
         }
 
-        BsonContainer(BsonContainer&& v) {
-            m_type = v.m_type;
-
-            if (m_type == DOCUMENT) {
-                document = v.document;
-                v.document = nullptr;
-            } else {
-                array = v.array;
-                v.array = nullptr;
-            }
+        explicit operator BsonArray&() noexcept {
+            REALM_ASSERT(is_array());
+            return *array.release();
         }
 
         bool is_array() const { return m_type == ARRAY; }
         bool is_document() const { return m_type == DOCUMENT; }
 
-        void push_back(std::pair<std::string, Bson> value) {
+        void push_back(const std::string& key, Bson&& value) {
             if (m_type == DOCUMENT) {
-                (*document)[value.first] = value.second;
+                document->operator[](key) = std::move(value);
```

Comment:
Explicitly calling operator[] is very unusual and the previous version of this code was more idiomatic.